packages/fast-development-site-react/src/components/site/index.tsx

```diff
@@ -93,87 +95,125 @@ class Site extends React.Component<ISiteProps & IManagedClasses<ISiteManagedClas
+                key={path}
+                path={path}
+                component={this.renderShellRow.bind(this, route)}
+            />
         );
     }
 
     private renderShellHeader(): JSX.Element {
         return (
             <ShellHeader>
-                {this.getSlotItems(this, ShellSlot.header)}
+                {this.getChildrenBySlot(this, ShellSlot.header)}
                 <span className={this.props.managedClasses.site_headerTitle}>
                     {this.props.title}
                 </span>
             </ShellHeader>
         );
     }
 
-    private renderShellRow(CanvasComponent: any, path: string): JSX.Element {
+    private renderShellRow(route: IComponentRoute): JSX.Element {
         return (
             <ShellRow>
                 <ShellPane collapsed={this.state.tableOfContentsCollapsed}>
                     {this.getPaneCollapseToggle()}
-                    {this.getSlotItems(this, ShellSlot.pane)}
+                    {this.getChildrenBySlot(this, ShellSlot.pane)}
                     <ul className={this.props.managedClasses.site_paneToc}>
-                        {this.getRootToc(this.props.children, SiteSlot.category, path, "/")}
+                        {this.getRootToc(this.props.children, SiteSlot.category, route.route, "/")}
                     </ul>
                 </ShellPane>
-                <ShellCanvas>
+                <ShellCanvas >
                     <ShellActionBar>
-                        {this.getSlotItems(this, ShellSlot.actionBar)}
+                        <ActionBar
+                            onComponentViewChange={this.onComponentViewChange}
+                            componentView={this.state.componentView}
+                        />
                     </ShellActionBar>
-                    {this.getSlotItems(this, ShellSlot.canvas)}
-                    {CanvasComponent}
+                    <ComponentView>
+                        {this.getChildrenBySlot(this, ShellSlot.canvas)}
+                        {route.component}
+                    </ComponentView>
                 </ShellCanvas>
             </ShellRow>
         );
     }
 
+    private onComponentViewChange = (type: ComponentViewTypes): void => {
+        this.setState({
+            componentView: type
+        });
+    }
+
     private renderShellInfoBar(): JSX.Element {
         return (
             <ShellInfoBar>
-                {this.getSlotItems(this, ShellSlot.infoBar)}
+                {this.getChildrenBySlot(this, ShellSlot.infoBar)}
             </ShellInfoBar>
         );
     }
 
     /**
      * Gets all of the potential routes as strings to be used to build the shell
      */
-    private getRoutes(items: JSX.Element, baseRoute: string, slot: string, routes?: IComponentRoutes[]): IComponentRoutes[] {
-        let currentRoutes: IComponentRoutes[] = routes;
+    private getRoutes(items: JSX.Element, baseRoute: string, slot: SiteSlot, routes?: IComponentRoute[]): IComponentRoute[] {
```

Comment:
Do we need to check that routes exist here?

Reply:
na, it is handled if it is undefined